gnomad/utils/constraint.py

```diff
@@ -360,3 +412,285 @@ def collapse_strand(
+    )
+
+    if pops:
+        # Map the models to their corresponding populations.
+        pop_models_arr = plateau_models["pop"]
+        for idx, pop in enumerate(pops):
+            plateau_models[pop] = pop_models_arr[idx]
+        plateau_models.pop("pop")
+    plateau_models = hl.struct(**plateau_models)
+
+    # Filter to sites with coverage below `cov_cutoff` and larger than 0.
+    low_cov_ht = coverage_ht.filter(
+        (coverage_ht.exome_coverage < cov_cutoff) & (coverage_ht.exome_coverage > 0)
+    )
+
+    # Metric that represents the relative mutability of the exome calculated on high
+    # coverage sites and will be used as scaling factor when building the coverage
+    # model.
+    high_coverage_scale_factor = high_cov_ht.aggregate(
+        hl.agg.sum(high_cov_ht.observed_variants)
+        / hl.agg.sum(high_cov_ht.possible_variants * high_cov_ht.mu_snp)
+    )
+
+    # Generate a Table with all necessary annotations (x and y listed above)
+    # for the coverage model.
+    low_cov_group_ht = low_cov_ht.group_by(
+        log_coverage=hl.log10(low_cov_ht.exome_coverage)
+    ).aggregate(
+        low_coverage_oe=hl.agg.sum(low_cov_ht.observed_variants)
+        / (
+            high_coverage_scale_factor
+            * hl.agg.sum(low_cov_ht.possible_variants * low_cov_ht.mu_snp)
+        )
+    )
+
+    # Build the coverage model.
+    # TODO: consider weighting here as well
+    coverage_model_expr = build_coverage_model(
+        low_coverage_oe_expr=low_cov_group_ht.low_coverage_oe,
+        log_coverage_expr=low_cov_group_ht.log_coverage,
+    )
+    coverage_model = tuple(low_cov_group_ht.aggregate(coverage_model_expr).beta)
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
+    cpg_expr: hl.expr.BooleanExpression,
+    mu_snp_expr: hl.expr.Float64Expression,
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_exprs_arr: Dict[str, hl.ArrayNumericExpression] = None,
+    weighted: bool = False,
+) -> Dict[str, Union[Dict[bool, hl.expr.ArrayExpression], hl.ArrayExpression]]:
+    """
+    Build plateau models to calibrate mutation rate to compute predicted proportion observed value.
+
+    The x and y of the plateau models:
+    - x: `mu_snp` - mutation rate
+    - y: proportion observed ('observed_variants' or 'observed_{pop}' / 'possible_variants')
+
+    :param cpg_expr: BooleanExpression noting whether a site is a CPG site.
+    :param mu_snp_expr: Float64Expression of the mutation rate.
+    :param observed_variants_expr: Int64Expression of the observed variant counts
+      for each combination of keys in `ht`.
+    :param possible_variants_expr: Int64Expression of the possible variant counts
+      for each combination of keys in `ht`.
+    :param pop_observed_variants_exprs_arr: ArrayExpression includes all observed
+      variant counts ArrayNumericExpressions for specified populations. Default is None.
```

Comment:
```suggestion
    :param pop_observed_variants_exprs_arr: Nested ArrayExpression with all observed
      variant counts ArrayNumericExpressions for specified populations. Default is None.
```
slightly clearer. maybe it would be clearer to just show an example, like `[[0, 0, 1], [1, 1, 2]]`

Reply:
looking at this again, this might be a very wide array. I guess we'll cross that bridge if we get to it

Reply:
Just to confirm. Is it ok to have such wide array?

Reply:
we won't know until it causes issues, though you could ask Tim?

Reply:
I see. It works fine on v2 but may raise error on v4. It's Tim's suggestion to pass in an ArrayExpression instead of dictionary. I guess we will see when testing the pipeline on v4!